### PR TITLE
some modification in `matmul` function

### PR DESCRIPTION
**Pull request by vtavana**
In this PR, a few minor changes are made in `matmul` function:

1) a TODO is removed since now that `dpnp.are_same_logical_tensors` check is added to `get_result_array` function.
2) maximum batch size that can be used in MKL function, is updated. The current value is validated to work on Intel Iris Xe and PVC GPUs.
3) The `usm_type` of out keyword in matmul function does not need to match the `usm_type` of input arrays, the corresponding check is removed.
4) minor changes are made to organize the code

- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 1]
```diff
@@ -364,7 +366,6 @@ def _gemm_batch_matmul(exec_q, x1, x2, res):
         )
         _manager.add_event_pair(ht_ev, blas_ev)
 
-    res_shape = res.shape
     _, res_is_c_contig, res_is_f_contig = _define_contig_flag(res)
     if row_major:
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:370]: Do we need to call `return dpnp.ascontiguousarray(res)` below here, keeping in mind use case when `out` memory is used?
Or probably we need to move that call outside `_gemm_batch_matmul`, for instance somewhere to the end of `dpnp_matmul`?

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 2]
```diff
@@ -150,7 +150,6 @@ def _create_result_array(
         if (
```
[the remark this answers is not in the document]
Comment (vtavana) [reply dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:150]: 1) For calling `gemm_batch`, **each 2D array** should be c-contiguous or f-contiguous. Before calling `gemm_batch`, we reshape the input array to a 3D array (batch of 2D arrays). Each of these 2-D array in a c-contiguous 3D array is also c-contiguous. However, each of these 2-D array in a f-contiguous 3D array is **not** f-contiguous.

```
>>> a = np.ones((3, 4, 5), order="C")
>>> a[0,:,:].flags # first array in the batch is also c-contig
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  
  
>>> a = np.ones((3, 4, 5), order="F")
>>> a[0,:,:].flags # first array in the batch is not f-contig
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
```

With this expalantion, it becomes clear that we cannot use f-contiguous arrays in `gemm_batch`. So, when column major version of `gemm_batch` is useful? It is useful for some special cases like below example where we have a bach of 2-D f-contiguous arrays.
```
>>> a=np.ones((2,3,4))
>>> a=a.transpose(0, 2, 1)
>>> a[0,:,:].flags # first array in the batch is f-contig
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
```

2) As mentioned, before calling `gemm_batch`, inputs array are reshaped to 3D array.
Reshaping of a 3D array does not change the order. The returned array is always a view of the given array.
```
>>> a = np.ones((3, 4, 5), order="F")
>>> b = a.reshape(-1, a.shape[-2], a.shape[-1])
>>> b.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
```
But reshaping a 4D array will returns a c-contiguous array.
For this case, if the given array is c-contiguous, returned array is a view and if the given array is F-contiguous the returned array is a copy.
```
>>> a = np.ones((2, 3, 4, 5), order="F")
>>> b = a.reshape(-1, a.shape[-2], a.shape[-1])
>>> b.flags # b is c-contig and a copy of a
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
```
3) In your example, your input array is 4D and f-contiguous, so when you remove `contig_flag` check, during the reshape operation to a 3D array, a copy is generated.
Alternatively, with having `contig_flag` check, a copy is created in `_create_result_array` function. 
So, for this case, there is a copy operation regardless of including/excluding `contig_flag` check.
However, if input array is 3D, it is important to have `contig_flag` check. Without `contig_flag` check, the `out` remains f-contiguous which is an invalid option due to item 1 described above.

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 3]
```diff
@@ -325,10 +324,13 @@ def _get_result_shape(x1, x2, out, np_flag):
...
     # arrays here are already at least 3D, make them 3D
-    x1 = dpnp.reshape(x1, (-1, x1.shape[-2], x1.shape[-1]))
-    x2 = dpnp.reshape(x2, (-1, x2.shape[-2], x2.shape[-1]))
+    x1_shape = x1.shape
+    x2_shape = x2.shape
+    x1 = dpnp.reshape(x1, (-1, x1_shape[-2], x1_shape[-1]))
+    x2 = dpnp.reshape(x2, (-1, x2_shape[-2], x2_shape[-1]))
     orig_shape = res.shape
-    res = dpnp.reshape(res, (-1, res.shape[-2], res.shape[-1]))
+    res = dpnp.reshape(res, (-1, orig_shape[-2], orig_shape[-1]))
```
[the remark this answers is not in the document]
Comment (vtavana) [reply dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:332]: For calling gemm_batch, each 2D array should be c-contiguous or f-contiguous. For an N-D and f-contiguous array, the 2-D arrays are not c- nor f-contiguous so it is and invalid option. I will add check point to the backend to throw an error indicating that the each 2D array should be c-contiguous or f-contiguous.

**Review by antonwolfy — APPROVED** [review]
Thank you @vtavana for the improvements implements and for the great explanation in review comments. The behavior is much more clear now for me.

**Pull request merged**